src/facet.jl

```diff
@@ -1,126 +1,306 @@
+    hideinnerdecorations!(aes, attr.hidexdecorations, attr.hideydecorations)
+
+    # delete empty axes
+    deleteemptyaxes!(aes)
+
+    # add facet labels
+    panel_labels!(fig, aes, scale)
+
+    # span axis labels if appropriate
+    nonempty_aes = get_nonempty_aes(aes)
+    is2d = isaxis2d(first(nonempty_aes))
+
+    if is2d && consistent_ylabels(aes)
+        span_ylabel!(fig, aes)
+    end
+    if is2d && consistent_xlabels(aes)
+        span_xlabel!(fig, aes)
     end
+    
     return
 end
 
-# F. Greimel implementation from AlgebraOfGraphics
-
-# TODO: add configuration options here (esp. to determine when / how to link)
-function facet_grid!(fig, aes::AbstractMatrix{AxisEntries})
+function facet_grid!(fig, aes::AbstractMatrix{AxisEntries}; facet)
     M, N = size(aes)
     row_scale, col_scale = map(sym -> get(aes[1].scales, sym, nothing), (:row, :col))
     all(isnothing, (row_scale, col_scale)) && return
-    hideinnerdecorations!(aes)
-    linkaxes!(aes...)
 
-    nonempty_aes = filter(ae -> !isempty(ae.entries), aes)
+    # Link axes and hide decorations if appropriate
+    attr = clean_facet_attributes(aes, facet)
+    link_axes!(aes; attr.linkxaxes, attr.linkyaxes)
+    hideinnerdecorations!(aes, attr.hidexdecorations, attr.hideydecorations)
 
-    ax = first(nonempty_aes).axis
+    # span axis labels if appropriate
+    nonempty_aes = get_nonempty_aes(aes)
+    is2d = isaxis2d(first(nonempty_aes))
+
+    if !isnothing(row_scale) && consistent_ylabels(aes)
+        is2d && span_ylabel!(fig, aes)
+        row_labels!(fig, aes, row_scale)
+    end
+    if !isnothing(col_scale) && consistent_xlabels(aes)
+        is2d && span_xlabel!(fig, aes)
+        col_labels!(fig, aes, col_scale)
+    end
+    return
+end
+
+function clean_facet_attributes(aes, facet)
+    linkxaxes = get(facet, :linkxaxes, automatic)
+    linkyaxes = get(facet, :linkyaxes, automatic)
+    hidexdecorations = get(facet, :hidexdecorations, automatic)
+    hideydecorations = get(facet, :hideydecorations, automatic)
+
+    if linkxaxes ∉ [:all, :colwise, :none, true, false, automatic]
+        @warn "Replaced invalid keyword linkxaxes = $linkxaxes by automatic"
+        linkxaxes = automatic
+    end
+
+    if linkyaxes ∉ [:all, :rowwise, :none, :true, false, automatic] 
+        @warn "Replaced invalid keyword linkyaxes = $linkyaxes by automatic"
+        linkyaxes = automatic
+    end
+
+    if hidexdecorations ∉ [true, false, automatic]
+        @warn "Replaced invalid keyword hidexdecorations = $hidexdecorations by automatic"
+        hidexdecorations = automatic
+    end
+
+    if hideydecorations ∉ [true, false, automatic] 
+        @warn "Replaced invalid keyword hideydecorations = $hideydecorations by automatic"
+        hideydecorations = automatic
+    end
+
+    if linkxaxes ∈ [automatic, true]
+        if colwise_consistent_xlabels(aes)
+            linkxaxes = :colwise
+		 else
+         	linkxaxes = :none
+		 end
```

Comment:
Indentation issues here as well, shouldn't matter though as I imagine this code will not be needed if we switch to linking `:all` by default. I mentioned it in the discussion but I don't really like that `true` is weaker than `:all`, I think they should be equal.

We could have a separate symbol (eg `:minimal`) to mean: "only link axes if that allows to hide decorations" (which is what happens in this conditional AFAIU).